[AMDGPU] Extend __builtin_amdgcn_ds_bpermute argument types

yxsamliu opened the pull request:
This change makes __builtin_amdgcn_ds_bpermute polymorphic over arbitrary trivially copyable types: the source (second) argument may be any such type, and the return type matches it. CodeGen treats values as one or more 32-bit words, permuting each word (and packing/unpacking any tail bytes) in a byte-wise manner; integers honor signedness, pointers use ptrtoint/inttoptr, and non-integers are bitcast per DataLayout. The implementation establishes a general CodeGen approach for byte-wise lane operations on arbitrary-sized and -typed Clang builtins, intended to be reusable for other byte-wise intrinsics.

This patch also introduces a new AMDGPUSupport.rst section documenting Clang's AMDGPU-specific builtins, starting with __builtin_amdgcn_ds_bpermute. The goal is to provide clear, centralized reference material for developers writing performance-critical GPU kernels as demand for AMDGPU builtin documentation grows; future updates will extend this section with additional builtins, their semantics, usage patterns, and target requirements.

shiltian commented on clang/lib/CodeGen/TargetBuiltins/AMDGPU.cpp:
```diff
@@ -159,6 +159,119 @@ Value *EmitAMDGPUGridSize(CodeGenFunction &CGF, unsigned Index) {
 }
+// Lowers __builtin_amdgcn_ds_bpermute to the corresponding LLVM intrinsic with
+// careful bit-level coercions of operands and result to match Clang types.
+llvm::Value *emitAMDGCNDsBpermute(clang::CodeGen::CodeGenFunction &Cgf,
```
I think `CGF` and `CGM` are widely used in the front end.

yxsamliu replied:
will fix

shiltian commented on clang/lib/CodeGen/TargetBuiltins/AMDGPU.cpp:
```diff
@@ -159,6 +159,119 @@ Value *EmitAMDGPUGridSize(CodeGenFunction &CGF, unsigned Index) {
+
+    if (Ty->isIntegerTy()) {
+      unsigned BitWidth = Ty->getIntegerBitWidth();
+      if (BitWidth < 32) {
```
maybe `<=`?

yxsamliu commented on clang/lib/Sema/SemaAMDGPU.cpp:
```diff
@@ -18,6 +18,132 @@
+
+  // Iterate over class conversion operators and pick the first that yields a
+  // 32-bit type.
+  for (auto *MethodDecl : RecordDecl->methods()) {
```
will fix

yxsamliu commented on clang/lib/CodeGen/TargetBuiltins/AMDGPU.cpp:
```diff
@@ -159,6 +159,119 @@ Value *EmitAMDGPUGridSize(CodeGenFunction &CGF, unsigned Index) {
+  llvm::Type *SourceTy = Source->getType();
+
+  if (SourceTy->isDoubleTy()) {
+    llvm::Value *AsFloat = Builder.CreateFPTrunc(Source, Builder.getFloatTy());
```
we also accept 64 bit integer types. They are all implicitly converted to 32 bit types (double to float, long to int). The situation is similar to the situation of double passed to a function accepting float, or long passed to a function accepting int, which causes implicit conversion with a warning.

rampitec replied:
It would be useful to accept double and long though and produce 2 instructions.

rampitec reviewed (commented):
Add some tests with complex and pointers?

b-sumner commented on clang/docs/AMDGPUSupport.rst:
```diff
@@ -61,3 +61,111 @@ Predefined Macros
+Conversion rules:
+- Builtins <= 32 bits (e.g., ``char``, ``short``, ``int``, ``uint32_t``, ``float``): bitcast to/from i32.
+- ``double``: convert to ``float`` before permutation; convert back to ``double`` after (may warn).
+- Long integers (> 32 bits, e.g., ``long long``, ``__int128``): convert to i32 before permutation; convert back after (may warn).
```
Conversion of > 32 bit types to 32-bits destroys information and is unwanted.  The other communication functions do not do this.  When the type is larger than 32-bits the idea is to split the data into 32-bit chunks and handle each separately.

yxsamliu committed:
[AMDGPU] Extend __builtin_amdgcn_ds_bpermute argument types

This change makes __builtin_amdgcn_ds_bpermute polymorphic over arbitrary trivially copyable types: the source (second) argument may be any such type, and the return type matches it. CodeGen converts the source value to an integer of the same bit width, zero-extends it to the next multiple of 32 bits, and permutes each 32-bit word independently via the ds_bpermute_b32 intrinsic. The result is reassembled, truncated back to the original bit width, and converted back to the source type. Scalars and vectors use bitcast/ptrtoint; aggregates use a store-as-aggregate + load-as-integer pattern that SROA can optimize away.

This patch also introduces a new AMDGPUSupport.rst section documenting Clang's AMDGPU-specific builtins, starting with __builtin_amdgcn_ds_bpermute.

arsenm reviewed (commented):
I can understand handling the other primitive types, like short2 or a pointer, but I think it's unreasonable for this builtin to support all of these aggregates

yxsamliu commented on clang/lib/CodeGen/TargetBuiltins/AMDGPU.cpp:
```diff
@@ -159,6 +159,242 @@ Value *EmitAMDGPUGridSize(CodeGenFunction &CGF, unsigned Index) {
+
+  // Size/bitwidth and coercion helpers for arbitrary first-class types.
+  auto GetBitWidth = [&](llvm::Type *Ty) -> unsigned {
+    return DL.getTypeSizeInBits(Ty).getFixedValue();
```
removed

shiltian commented:
I think what you proposed can be easily implemented as a function in a header, rather than as a compiler builtin. I also don't think this falls within the scope of a compiler builtin. You can probably just implement the logic in one of the compiler internal headers.

arsenm commented on clang/test/SemaHIP/builtin-amdgcn-ds-bpermute.hip:
```diff
@@ -0,0 +1,37 @@
+typedef short short2 __attribute__((ext_vector_type(2)));
+typedef int int2 __attribute__((ext_vector_type(2)));
+
+struct A { int a; };
```
Struct should not be allowed

shiltian commented:
On a side note, can you add documentation? The infrastructure has already been merged and we have the website already.
